confounds/metrics.py

```diff
@@ -6,3 +6,102 @@
 degree of harmonization achieved (e.g. reduction in variance of means/medians)
 
 """
+
+import numpy as np
+
+from confounds import Residualize
+from scipy import stats
+
+
+def partial_correlation(X, y=None):
+    """
+    Calculates the pairwise partial correlations between all of the variables in X with respect to some confounding variables y
+
+    Can be used as a measure of the strength of the relationship between two variables of interest while controlling for some other variables.
+
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            The training input samples.
+    y : {array-like, sparse matrix}, shape (n_samples, n_covariates)
+            This does not refer to target as is typical in scikit-learn.
+
+    Returns
+    -------
+    partial_correlations : ndarray
+        Returns the pairwise partial correlations of each variable in X
+    """
+    resx = Residualize()
+    resx.fit(X, y)
+    deconfound_X = resx.transform(X, y)
+    return np.corrcoef(deconfound_X, rowvar=False)
+
+
+def partial_correlation_t_test(X, y=None):
+    """
+    Calculates the t-statistic and p-value for pairwise partial correlations between all of the variables in X with
+    respect to some confounding variables y.
+
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            The training input samples.
+    y : {array-like, sparse matrix}, shape (n_samples, n_covariates)
+            This does not refer to target as is typical in scikit-learn.
+
+    Returns
+    -------
+    corr_p : ndarray
+        Returns the pairwise partial correlations of each variable in X
+    t_statistic : ndarray
+        Returns the associated t-statistics for these pairwise partial correlations
+    statistical_significance : ndarray
+        Returns the associated p-values for these pairwise partial correlations
+    """
+    corr_p = partial_correlation(X, y=y)
+    n = X.shape[0]
+    g = y.shape[1]
+    # Replace perfect correlations to ensure large but not infinite t statistic
+    corr_p[corr_p == 1] = 1 - 1e-7
+    # partial correlation degrees of freedom
+    df = n - 2 - g
+    t_statistic = corr_p * np.sqrt(df / (1 - corr_p ** 2))
```

Comment:
is there a reference/citation you base this formula on? Can we mention it in the docs for this function?

Reply:
Dinga et al. make this really clear but it's essentially public domain. Have added the Dinga reference to this code section

Reply:
Sure! but it does help to summarize your understanding in few sentences so users are not redirected! scikit-learn docs are great in that way as they tend more or less self-sufficient when possible. Do what you can.